### PR TITLE
RavenDB-15538: Replicate all document attachments if there are MissingAttachmentsInLastBatch

### PR DESCRIPTION
**Pull request by garayx**

### CODE REVIEW

File: src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs [hunk 1]
```diff
@@ -379,6 +346,66 @@ public bool ExecuteReplicationOnce(TcpConnectionOptions tcpConnectionOptions, Ou
             }
         }
 
+        private bool CanContinueBatch(HashSet<short> missingTx, TimeSpan delay, ReplicationBatchItem item, long currentNext, long size, DocumentsOperationContext documentsContext,
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs]: Can we make a class out most of the args here? like `ReplicationState`

File: src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs [hunk 2]
```diff
@@ -379,6 +346,66 @@ public bool ExecuteReplicationOnce(TcpConnectionOptions tcpConnectionOptions, Ou
...
+                missingTx.Remove(lastTransactionMarker);
+
+                if (missingTx.Count != 0)
+                {
+                    return true;
+                }
+            }
+
+            var batchSize = _parent._database.Configuration.Replication.MaxItemsCount;
+            var maxSizeToSend = _parent._database.Configuration.Replication.MaxSizeToSend;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs]: And maybe put those in that class.

**Review by karmeli87 — APPROVED** [review]

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs [hunk 3]
```diff
@@ -379,6 +356,62 @@ public bool ExecuteReplicationOnce(TcpConnectionOptions tcpConnectionOptions, Ou
             }
         }
 
+        private bool CanContinueBatch(ReplicationState state, ref long next)
+        {
+            if (MissingAttachmentsInLastBatch)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs:361]: Isn't this called for any item? In that case, what is the meaning of removing it one at a time?

Comment (garayx) [reply src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs:361]: this code called only when `item.TransactionMarker != replicationState.LastTransactionMarker`
then we try to remove `replicationState.LastTransactionMarker`, if there are `state.MissingTxMarkers` we have to continue the replication batch, else we allow to interrupt it if needed

**Pull request merged**